File: mvn-advent-of-code/src/main/java/advent/twenty_fifteen/Day16AuntSue.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;

import java.util.*;

public class Day16AuntSue implements DailyProblem<Integer, Integer> {

    // Sue 464: pomeranians: 8, cars: 5, vizslas: 0
    private static Map<String, Integer> PARSE(String line) {
        Map<String, Integer> result = new HashMap<>();

        int index = line.indexOf(":");
        result.put("INDEX", Integer.parseInt(line.substring(3, index).trim()));
        String[] bits = line.substring(index + 1).split(",");
        for (String bit : bits) {
            String[] chunks = bit.split(":");
            result.put(chunks[0].trim(), Integer.parseInt(chunks[1].trim()));
        }

        result.put("SCORE 1", calculateScoreOne(result));
        result.put("SCORE 2", calculateScoreTwo(result));

        return result;
    }

    private static int calculateScoreOne(Map<String, Integer> value) {
        int score = 0;

        if (Integer.valueOf(3).equals(value.get("children"))) score++;
        if (Integer.valueOf(7).equals(value.get("cats"))) score++;
        if (Integer.valueOf(2).equals(value.get("samoyeds"))) score++;
        if (Integer.valueOf(3).equals(value.get("pomeranians"))) score++;
        if (Integer.valueOf(0).equals(value.get("akitas"))) score++;
        if (Integer.valueOf(0).equals(value.get("vizslas"))) score++;
        if (Integer.valueOf(5).equals(value.get("goldfish"))) score++;
        if (Integer.valueOf(3).equals(value.get("trees"))) score++;
        if (Integer.valueOf(2).equals(value.get("cars"))) score++;
        if (Integer.valueOf(1).equals(value.get("perfumes"))) score++;

        return score;
    }

    private static int calculateScoreTwo(Map<String, Integer> value) {
        int score = 0;

        if (Integer.valueOf(3).equals(value.get("children"))) score++;
        if (Integer.valueOf(2).equals(value.get("samoyeds"))) score++;
        if (Integer.valueOf(0).equals(value.get("akitas"))) score++;
        if (Integer.valueOf(0).equals(value.get("vizslas"))) score++;
        if (Integer.valueOf(2).equals(value.get("cars"))) score++;
        if (Integer.valueOf(1).equals(value.get("perfumes"))) score++;

        if (value.containsKey("cats") && Integer.valueOf(7).compareTo(value.get("cats")) <= 0) score++;
        if (value.containsKey("trees") && Integer.valueOf(3).compareTo(value.get("trees")) <= 0) score++;

        if (value.containsKey("pomeranians") && Integer.valueOf(3).compareTo(value.get("pomeranians")) >= 0) score++;
        if (value.containsKey("goldfish") && Integer.valueOf(5).compareTo(value.get("goldfish")) >= 0) score++;

        return score;
    }

    private final int part1Answer;
    private final int part2Answer;

    public Day16AuntSue(String filename) {
        List<Map<String, Integer>> lines = FileUtilities.readLines(filename, Day16AuntSue::PARSE);

        lines.sort(Comparator.comparingLong(o -> o.get("SCORE 1")));
        Collections.reverse(lines);

        this.part1Answer = lines.get(0).get("INDEX");

        lines.sort(Comparator.comparingLong(o -> o.get("SCORE 2")));
        Collections.reverse(lines);

        this.part2Answer = lines.get(0).get("INDEX");
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day16AuntSue.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

case class AuntSue(id: Int, attributes: Map[String, Int])

class Day16AuntSue(filename: String) extends DailyProblem[Int, Int] {

  private val tickerTape = Map("children" -> 3, "cats" -> 7, "samoyeds" -> 2, "pomeranians" -> 3, "akitas" -> 0, "vizslas" -> 0, "goldfish" -> 5, "trees" -> 3, "cars" -> 2, "perfumes" -> 1)

  private val pattern = "Sue ([0-9]+): ([A-Za-z]+): ([0-9]+), ([A-Za-z]+): ([0-9]+), ([A-Za-z]+): ([0-9]+)".r

  private def parse(line: String): AuntSue = {
    line match {
      case pattern(id, a, b, c, d, e, f) => AuntSue(id.toInt, Map(a -> b.toInt, c -> d.toInt, e -> f.toInt))
    }
  }

  private def score1(sue: Map[String, Int]): Int = {
    sue.map(entry => if (tickerTape(entry._1) == entry._2) 1 else 0).sum
  }

  private def score2(sue: Map[String, Int]): Int = {
    sue.map(entry =>
      entry._1 match {
        case "cats" | "trees" => if (tickerTape(entry._1) <= entry._2) 1 else 0
        case "pomeranians" | "goldfish" => if (tickerTape(entry._1) >= entry._2) 1 else 0
        case _ => if (tickerTape(entry._1) == entry._2) 1 else 0
      }).sum
  }

  private val sues = FileUtilities.readFile(filename, parse)

  override val part1Answer: Int = sues.map(sue => (sue.id, score1(sue.attributes))).sortWith((a, b) => a._2 > b._2).map(a => a._1).head
  override val part2Answer: Int = sues.map(sue => (sue.id, score2(sue.attributes))).sortWith((a, b) => a._2 > b._2).map(a => a._1).head
}
